File: THCS/2023_2024/NgheAn_2324/Cau2/B2.py
from itertools import combinations

# Đọc dữ liệu từ tệp B2.INP
with open("B2.INP", "r") as file:
    t = int(file.readline().strip())
    tests = [file.readline().strip().split() for _ in range(t)]

# Hàm để kiểm tra số có chia hết cho 5
def divisible_by_5(number_str):
    return int(number_str) % 5 == 0

# Hàm để kiểm tra số có chia hết cho 3
def divisible_by_3(number_str):
    return int(number_str) % 3 == 0

results = []
for number_str, flag in tests:
    n = len(number_str)
    best_number = '-1'  # Khởi tạo với giá trị rất nhỏ
    
    # Tạo các chuỗi con theo thứ tự từ lớn nhất đến nhỏ nhất
    for length in range(n, 0, -1):
        # Tạo tất cả các tổ hợp có độ dài là 'length'
        for combo in combinations(number_str, length):
            sub_number = ''.join(combo)
            if flag == '0' and divisible_by_5(sub_number):
                # Chia hết cho 5
                if best_number == '-1' or int(sub_number) > int(best_number):
                    best_number = sub_number
            elif flag == '1' and divisible_by_3(sub_number):
                # Chia hết cho 3
                if best_number == '-1' or int(sub_number) > int(best_number):
                    best_number = sub_number
        if best_number != '-1':  # Nếu tìm thấy kết quả
            break
    
    results.append(best_number)

# Ghi kết quả vào tệp B2.OUT
with open("B2.OUT", "w") as file:
    for result in results:
        file.write(result + "\n")

File: THCS/2023_2024/NgheAn_2324/Cau2/B2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

// Kiểm tra số chia hết cho 5
bool divisible_by_5(const string& number_str) {
    return stoi(number_str) % 5 == 0;
}

// Kiểm tra số chia hết cho 3
bool divisible_by_3(const string& number_str) {
    int sum = 0;
    for (char c : number_str) {
        sum += (c - '0');
    }
    return sum % 3 == 0;
}

int main() {
    ifstream infile("B2.INP");
    ofstream outfile("B2.OUT");

    int t;
    infile >> t;
    vector<string> results;

    for (int i = 0; i < t; ++i) {
        string number_str;
        int flag;
        infile >> number_str >> flag;

        string best_number = "-1";

        int n = number_str.size();

        for (int length = n; length > 0; --length) {
            // Tạo các chuỗi con với độ dài là 'length'
            for (int start = 0; start <= n - length; ++start) {
                string sub_number = number_str.substr(start, length);

                // Kiểm tra điều kiện
                if (flag == 0 && divisible_by_5(sub_number)) {
                    if (best_number == "-1" || stoll(sub_number) > stoll(best_number)) {
                        best_number = sub_number;
                    }
                } else if (flag == 1 && divisible_by_3(sub_number)) {
                    if (best_number == "-1" || stoll(sub_number) > stoll(best_number)) {
                        best_number = sub_number;
                    }
                }
            }
            if (best_number != "-1") {
                break;
            }
        }
        results.push_back(best_number);
    }

    for (const string& result : results) {
        outfile << result << endl;
    }

    infile.close();
    outfile.close();
    return 0;
}
